README.md
# Readme for LiteOSCParser v1.4.0

This library parses and constructs OSC messages. It is designed to minimize
memory usage by providing a way to pre-allocate the internal buffers.

Bundles are supported via a simple container class.

## Features

Notable features of this library:

1. Ability to control memory usage.
2. Lightweight API.

## How to use

The classes you'll need are in the `qindesign::osc` namespace:
`LiteOSCParser` and possibly `OSCBundle`. A complete example of how to use
the parser is in the `AllInOne` example.

The main class documentation can be found in `src/LiteOSCParser.h`.

OSC bundles are managed using the simple `OSCBundle` container class.

## Installing as an Arduino library

Not all the files in this project are necessary in an installed library.
Only the following files and directories need to be there:

* `*.md`
* `LICENSE`
* `examples/`
* `library.json`
* `library.properties`
* `src/`

## Running the tests

There are tests included in this project that rely on a project called
[ArduinoUnit](https://github.com/mmurdoch/arduinounit).

Note that the code for ArduinoUnit is not included in this library and needs
to be downloaded separately.

## Notes on use

### Retrieving values

By default, if a value does not exist at a given index, a default value will be
returned. For example, `getFloat(2)` will try to return a 32-bit float value at
index 2. If the index is out of range or if the value type at that index is not
a float then a value of 0.0f will be returned.

To avoid this scenario, the following code is useful:

```c++
if (isFloat(2)) {
  float f = getFloat(2);
  // Do something with the value
}
```

Internally, `getFloat` also calls `isFloat` with the same index, and so
`isFloat` is actually called twice. To avoid two calls, there exist other getter
functions named `getIfXXX` (replacing _XXX_ with the appropriate type). For
example, `getIfFloat` to retrieve a float. These functions return a `bool`:
`true` if the value exists and was retrieved, and `false` otherwise.

For example:

```c++
float f;
if (getIfFloat(2, &f)) {
  // Do something with 'f'
}
```

## Code style

Code style for this project mostly follows the
[Google C++ Style Guide](https://google.github.io/styleguide/cppguide.html).

## References

1. [OSC 1.0 Specification](http://opensoundcontrol.org/spec-1_0)
2. [OSC 1.1 Specification](http://opensoundcontrol.org/spec-1_1)
3. [NIME 2009 paper](https://hangar.org/webnou/wp-content/uploads/2012/01/Nime09OSCfinal.pdf)
4. [TouchOSC](https://hexler.net/docs/touchosc)

---

Copyright (c) 2018-2019 Shawn Silverman


CHANGELOG.md
# Changelog for LiteOSCParser

This document details the changes between each release.

## [1.4.0]

### Added
* New `getIfXXX` functions that complement the `getXXX` functions by checking
  the type and possibly filling in the value in one step. These avoid two calls
  to `isXXX` when retrieving values: once by the calling code and once inside
  the `getXXX` functions.

### Changed
* Changed the licence to BSD-3-Clause-Clear.
* Updated the library description.

## [1.3.0]

### Changed
* Changed parameter types of the private `OSCBundle::add(const uint8_t *, int)`
  to `OSCBundle::add(const uint8_t *, int32_t)`.
* Tested with
  [ArduinoUnit v3.0.2](https://github.com/mmurdoch/arduinounit/releases/tag/v3.0.2).
* On systems where the size of type `double` is not 8 bytes,
  `LiteOSCParser::getDouble` will always return zero and
  `LiteOSCParser::addDouble` will always return `false`.

### Fixed
* Compiles on more platforms (listed using PlatformIO board names): `teensylc`,
  `teensy35`, `teensy36`, `megaatmega1280`, `megaatmega2560`, `uno`, `yun`,
  `featheresp32`, `huzzah`, `esp12e`, `nucleo_f302r8`, `nucleo_f303k8`,
  `nucleo_f103rb`. This is in addition to just compiling and testing on
  `teensy31`.
* Now making use of PlatformIO's `build_unflags` to ensure that the
  `-fsingle-precision-constant` does not take effect; the compiler was
  complaining that some float literals were still overflowing, even though
  `-fno-single-precision-constant` was being passed in `build_flags`.
* Changed `LiteOSCParser::addFloat` and `LiteOSCParser::getFloat` to use
  `static_assert`s to guarantee that the size of type `float` is 4 bytes. This
  ensures that memory is copied into a space having the correct size.


## [1.2.1]

### Added
* keywords.txt file.

## [1.2.0]

### Added
* OSC bundle parsing.

## [1.1.0]

### Added
* OSC bundle support, via a simple `OSCBundle` container class.
* A `LiteOSCParser::clear()` function that empties the message.
* A test for out-of-range matching.
* An all-in-one example.

### Changed

### Removed
* Default copy constructors are removed.

### Fixed
* The constructors now check if the initial memory allocation failed.
* The `match` function now returns the correct value if the offset is equal
  to the address length and the pattern to be matched is empty.

## LiteOSCParser v1.0.0

This is the first LiteOSCParser release. It provides a way to fix the amount
of memory being used internally. Bundles are not currently supported.

---

Copyright (c) 2018-2019 Shawn Silverman


src/tests.cpp
// tests.cpp is part of LiteOSCParser.
// (c) 2018 Shawn Silverman

// C++ includes
#include <cstdint>

// Other includes
#include <Arduino.h>
#include <ArduinoUnit.h>

// Project includes
#include "OSCParser.h"

OSCParser osc{64, 4};

#ifdef __cplusplus
extern "C" {
#endif
// Modified from:
// https://forum.pjrc.com/threads/186-Teensy-3-fault-handler-demonstration
void __attribute__((naked)) hard_fault_isr() {
  uint32_t* sp = 0;
  // this is from "Definitive Guide to the Cortex M3" pg 423
  asm volatile(
      "TST LR, #0x4\n\t"   // Test EXC_RETURN number in LR bit 2
      "ITE EQ\n\t"         // if zero (equal) then
      "MRSEQ %0, MSP\n\t"  //   Main Stack was used, put MSP in sp
      "MRSNE %0, PSP\n\t"  // else Process stack was used, put PSP in sp
      : "=r"(sp)
      :
      : "cc");

  Serial.print("!!!! Hard Crashed at pc=0x");
  Serial.print(sp[6], 16);
  Serial.print(", lr=0x");
  Serial.print(sp[5], 16);
  Serial.println(".");
  Serial.flush();

  while (true) {
    asm volatile("WFI"  // Wait For Interrupt.
    );
  }
}
#ifdef __cplusplus
}  // extern "C"
#endif

// The tests
#include "tests/add_args.inc"
#include "tests/address.inc"
#include "tests/args.inc"
#include "tests/match.inc"
#include "tests/memory.inc"
#include "tests/packet.inc"

void runTests() {
  Test::run();
}


src_tests/tests.cpp
// tests.cpp is part of LiteOSCParser.
// (c) 2018-2019 Shawn Silverman

// C++ includes
#ifdef __has_include
#if __has_include(<cstdint>)
#include <cstdint>
#else
#include <stdint.h>
#endif
#else
#include <cstdint>
#endif

// Other includes
#include <Arduino.h>
#include <ArduinoUnit.h>

// Project includes
#include "LiteOSCParser.h"

::qindesign::osc::LiteOSCParser osc{64, 4};

#if defined(CORE_TEENSY) && !defined(TEENSYLC)
#ifdef __cplusplus
extern "C" {
#endif
// Modified from:
// https://forum.pjrc.com/threads/186-Teensy-3-fault-handler-demonstration
void __attribute__((naked)) hard_fault_isr() {
  uint32_t* sp = 0;
  // this is from "Definitive Guide to the Cortex M3" pg 423
  asm volatile(
      "TST LR, #0x4\n\t"   // Test EXC_RETURN number in LR bit 2
      "ITE EQ\n\t"         // if zero (equal) then
      "MRSEQ %0, MSP\n\t"  //   Main Stack was used, put MSP in sp
      "MRSNE %0, PSP\n\t"  // else Process stack was used, put PSP in sp
      : "=r"(sp)
      :
      : "cc");

  Serial.print("!!!! Hard Crashed at pc=0x");
  Serial.print(sp[6], 16);
  Serial.print(", lr=0x");
  Serial.print(sp[5], 16);
  Serial.println(".");
  Serial.flush();

  while (true) {
    asm volatile("WFI"  // Wait For Interrupt.
    );
  }
}
#ifdef __cplusplus
}  // extern "C"
#endif
#endif  // CORE_TEENSY

// The tests
#include "tests/add_args.inc"
#include "tests/address.inc"
#include "tests/args.inc"
#include "tests/bundle.inc"
#include "tests/match.inc"
#include "tests/memory.inc"
#include "tests/packet.inc"

void setup() {
  Serial.begin(115200);
  while (!Serial && millis() < 4000) {
  }
  delay(4000);
}

void loop() {
  Test::run();
}


src/OSCBundle.cpp
// OSCBundle.cpp is part of LiteOSCParser.
// (c) 2018-2019 Shawn Silverman

#include "LiteOSCParser.h"

// C++ includes
#ifdef __has_include
#if __has_include(<cstdlib>)
#include <cstdlib>
#else
#include <stdlib.h>
#endif
#if __has_include(<cstring>)
#include <cstring>
#else
#include <string.h>
#endif
#else
#include <cstdlib>
#include <cstring>
#endif

namespace qindesign {
namespace osc {

OSCBundle::OSCBundle(int bufCapacity)
    : buf_(nullptr),
      bufSize_(0),
      bufCapacity_(0),
      dynamicBuf_(true),
      memoryErr_(false),
      isInitted_(false) {
  static_assert(sizeof(uint8_t) == 1, "sizeof(uint8_t) == 1");
  if (bufCapacity > 0) {
    dynamicBuf_ = false;
    if (bufCapacity < 16) {
      bufCapacity = 16;
    }
    buf_ = static_cast<uint8_t*>(malloc(bufCapacity));
    if (buf_ == nullptr) {
      memoryErr_ = true;
    } else {
      bufCapacity_ = bufCapacity;
    }
  }
}

OSCBundle::~OSCBundle() {
  if (buf_ != nullptr) {
    free(buf_);
  }
}

bool OSCBundle::init(uint64_t time) {
  if (!ensureCapacity(16)) {
    return false;
  }
  memcpy(buf_, "#bundle", 8);

  // Use 32-bit values
  uint32_t t = time >> 32;
  buf_[8] = t >> 24;
  buf_[9] = t >> 16;
  buf_[10] = t >> 8;
  buf_[11] = t;

  t = time;
  buf_[12] = t >> 24;
  buf_[13] = t >> 16;
  buf_[14] = t >> 8;
  buf_[15] = t;

  bufSize_ = 16;

  isInitted_ = true;
  return true;
}

bool OSCBundle::addMessage(const LiteOSCParser &osc) {
  return add(osc.getMessageBuf(), osc.getMessageSize());
}

bool OSCBundle::addBundle(const OSCBundle &bundle) {
  return add(bundle.buf(), bundle.size());
}

bool OSCBundle::parse(const uint8_t *buf, int32_t len) {
  if (len < 16 || (len & 0x03) != 0) {
    return false;
  }
  if (memcmp(buf, "#bundle", 8) != 0) {
    return false;
  }
  int index = 16;
  while (index < len) {
    int32_t size = static_cast<int32_t>(
        uint32_t{buf[index]} << 24 | uint32_t{buf[index + 1]} << 16 |
        uint32_t{buf[index + 2]} << 8 | uint32_t{buf[index + 3]});
    index += 4;
    if (size <= 0 || (size & 0x03) != 0 || index + size > len) {
      return false;
    }
    if (size >= 8 && memcmp(&buf[index], "#bundle", 8) == 0) {
      if (!parse(&buf[index], size)) {
        return false;
      }
    } else if (buf[index] != '/') {  // Rudimentary check for an OSC message
      return false;
    }
    index += size;
  }
  return true;
}

// --------------------------------------------------------------------------
//  Private functions
// --------------------------------------------------------------------------

bool OSCBundle::add(const uint8_t *buf, int32_t size) {
  if (!isInitted_) {
    return false;
  }
  if (!ensureCapacity(bufSize_ + 4 + size)) {
    return false;
  }

  uint32_t u = static_cast<uint32_t>(size);
  buf_[bufSize_++] = u >> 24;
  buf_[bufSize_++] = u >> 16;
  buf_[bufSize_++] = u >> 8;
  buf_[bufSize_++] = u;
  memcpy(&buf_[bufSize_], buf, size);
  bufSize_ += size;

  return true;
}

bool OSCBundle::ensureCapacity(int size) {
  if (size <= bufCapacity_) {
    return true;
  }
  if (!dynamicBuf_) {
    memoryErr_ = true;
    return false;
  }
  buf_ = static_cast<uint8_t*>(realloc(buf_, size));
  if (buf_ == nullptr) {
    memoryErr_ = true;
    return false;
  }
  bufCapacity_ = size;
  return true;
}

}  // namespace osc
}  // namespace qindesign


src/LiteOSCParser.h
// LiteOSCParser.h defines an OSC processor.
// This is part of LiteOSCParser.
// (c) 2018-2019 Shawn Silverman

#ifndef LITEOSCPARSER_H_
#define LITEOSCPARSER_H_

// C++ includes
#ifdef __has_include
#if __has_include(<cstdint>)
#include <cstdint>
#else
#include <stdint.h>
#endif
#else
#include <cstdint>
#endif

namespace qindesign {
namespace osc {

// LiteOSCParser parses and constructs OSC messages. The internal buffer
// and argument list can be either dynamically allocated or set to a
// specific size. Any functions that add to, initialize, or change the
// message return whether they were successful. One of the possible
// reasons for failure is that there's not enough space in one of the
// internal buffers, and more cannot be allocated. The isMemoryError()
// function can be used to determine this case.
//
// One strategy when creating a message would be to add everything needed
// and then check for a memory error afterwards, instead of after every
// addition.
//
// Some of the functions return a pointer to a spot in the internal buffer.
// Try not to mismanage those. :)
class LiteOSCParser {
 public:
  // Creates a new OSC parser. The maximum buffer size and argument count
  // are given. If the buffer size is non-positive then the buffer will be
  // dynamically allocated as needed. The size should be a multiple of four.
  //
  // Similarly, if the maximum argument count, maxArgCount, is non-positive,
  // then the internal array will be dynamically allocated as needed.
  // Otherwise, the maxmimum number of arguments in a message will be
  // limited to that count.
  //
  // The buffer size, bufSize, is given in bytes, and the maximum argument
  // count, maxArgCount, is given in ints, i.e. maxArgCount*sizeof(int).
  LiteOSCParser(int bufSize, int maxArgCount);

  // Initializes a new OSC parser having dynamic buffer and argument
  // allocation.
  LiteOSCParser() : LiteOSCParser(0, 0) {}

  // Not copyable
  LiteOSCParser(const LiteOSCParser &) = delete;
  LiteOSCParser &operator=(const LiteOSCParser &) = delete;

  ~LiteOSCParser();

  // ------------------------------------------------------------------------
  //  Creating
  // ------------------------------------------------------------------------

  // Initializes the message with a new address and no arguments. This
  // returns whether the initialization was successful. This will be
  // unsuccessful if the address does not start with a '/' or if there
  // isn't enough space in the internal buffer. The second condition can
  // be checked with a call to isMemoryError().
  bool init(const char *address);

  // Clears the message. Under the covers, this calls init with an
  // empty string and then ignores the result.
  void clear() {
    init("");
  }

  // Gets the total size of the encoded message.
  int getMessageSize() const {
    return bufSize_;
  }

  // Gets a pointer to the internal buffer, for sending the data elsewhere.
  // See getMessageSize() to retrieve the size.
  const uint8_t *getMessageBuf() const {
    return buf_;
  }

  // Adds a 32-bit int argument.
  bool addInt(int32_t i);

  // Adds a 32-bit float argument.
  bool addFloat(float f);

  // Adds a string.
  bool addString(const char *s);

  // Adds a blob having the given length.
  bool addBlob(const uint8_t *b, int len);

  // Adds a 64-bit long argument.
  bool addLong(int64_t h);

  // Adds a 64-bit time argument.
  bool addTime(uint64_t t);

  // Adds a 64-bit double argument.
  //
  // This will return `false` if the size of type `double` is not 8 bytes.
  bool addDouble(double d);

  // Adds a boolean.
  bool addBoolean(bool b);

  // Returns whether an insufficient buffer size is preventing the latest
  // message from being constructed.
  bool isMemoryError() const {
    return memoryErr_;
  }

  // ------------------------------------------------------------------------
  //  Parsing and matching
  // ------------------------------------------------------------------------

  // Parses the given buffer and returns whether the message is valid.
  // If this returns 'false' then isMemoryError() can be used to determine
  // whether the failure was due to not enough space in the internal buffer.
  bool parse(const uint8_t *buf, int len);

  // Returns whether the address fully matches the given pattern,
  // starting at offset in the address.
  bool fullMatch(int offset, const char *pattern) const;

  // Checks whether the address partially matches the given pattern and
  // returns an index one past the last character matched. This starts
  // matching at 'offset' in the address. This does not match partial
  // containers; for a match to happen, the next character of the address
  // must be a '/'.
  //
  // This will return -1 if offset is out of range and zero if there is
  // no match.
  int match(int offset, const char *pattern) const;

  // ------------------------------------------------------------------------
  //  Getters
  // ------------------------------------------------------------------------

  // Gets a pointer to the address. This returns a pointer into the
  // internal buffer.
  const char *getAddress() const;

  // Returns the current argument count.
  int getArgCount() const {
    if (tagsLen_ == 0) {
      return 0;
    }
    return tagsLen_ - 1;
  }

  // Returns the tag for the argument at the given index. This returns
  // '\0' if the index is out of range.
  char getTag(int index) const {
    if (index < 0 || tagsLen_ - 1 <= index) {
      return '\0';
    }
    return buf_[tagsIndex_ + index + 1];
  }

  // Checks if the argument at the given index is a 32-bit int.
  bool isInt(int index) const {
    return isTag(index, 'i');
  }

  // Checks if the argument at the given index is a 32-bit float.
  bool isFloat(int index) const {
    return isTag(index, 'f');
  }

  // Checks if the argument at the given index is a string.
  bool isString(int index) const {
    return isTag(index, 's');
  }

  // Checks if the argument at the given index is a blob.
  bool isBlob(int index) const {
    return isTag(index, 'b');
  }

  // Checks if the argument at the given index is a 64-bit long.
  bool isLong(int index) const {
    return isTag(index, 'h');
  }

  // Checks if the argument at the given index is an OSC-timetag.
  bool isTime(int index) const {
    return isTag(index, 't');
  }

  // Checks if the argument at the given index is a 64-bit double.
  bool isDouble(int index) const {
    return isTag(index, 'd');
  }

  // Checks if the argument at the given index is a char.
  bool isChar(int index) const {
    return isTag(index, 'c');
  }

  // Checks if the argument at the given index is a boolean.
  bool isBoolean(int index) const {
    return isTag(index, 'T') || isTag(index, 'F');
  }

  // Checks if the argument at the given index is a null.
  bool isNull(int index) const {
    return isTag(index, 'N');
  }

  // Checks if the argument at the given index is an impulse.
  bool isImpulse(int index) const {
    return isTag(index, 'I');
  }

  // Gets the 32-bit int at the given index. This will return zero
  // if the index is out of range or if the argument is the wrong type.
  int32_t getInt(int index) const;

  // Gets the value at the given index only if it exists and is a 32-bit int.
  // This will return `true` if the value was successfully retrieved and
  // `false` otherwise.
  //
  // This function is useful to avoid two calls to `isInt`, once by the calling
  // code, and a second time by `getInt`.
  bool getIfInt(int index, int32_t *v) const;

  // Gets the 32-bit float at the given index. This will return zero
  // if the index is out of range or if the argument is the wrong type.
  float getFloat(int index) const;

  // Gets the value at the given index only if it exists and is a 32-bit float.
  // This will return `true` if the value was successfully retrieved and
  // `false` otherwise.
  //
  // This function is useful to avoid two calls to `isFloat`, once by the
  // calling code, and a second time by `getFloat`.
  bool getIfFloat(int index, float *v) const;

  // Gets a pointer to the string stored at the given index. This returns
  // a pointer into the internal buffer, or nullptr if the index is out
  // of range or if the argument is the wrong type.
  const char *getString(int index) const;

  // Gets the blob length at the given index. This will return zero
  // if the index is out of range or if the argument is the wrong type.
  int getBlobLength(int index) const;

  // Gets a pointer to the stored stored at the given index. This returns
  // a pointer into the internal buffer, or nullptr if the index if out
  // of range or if the argument is the wrong type. See getBlobLength
  // to get the blob size.
  const uint8_t *getBlob(int index) const;

  // Gets the 64-bit long at the given index. This will return zero
  // if the index is out of range or if the argument is the wrong type.
  int64_t getLong(int index) const;

  // Gets the value at the given index only if it exists and is a 64-bit long.
  // This will return `true` if the value was successfully retrieved and
  // `false` otherwise.
  //
  // This function is useful to avoid two calls to `isLong`, once by the calling
  // code, and a second time by `getLong`.
  bool getIfLong(int index, int64_t *v) const;

  // Gets the 64-bit OSC-timetag at the given index. This will return zero
  // if the index is out of range or if the argument is the wrong type.
  //
  // The format is the representation used by NTP time stamps, where the
  // top 32 bits are the number of seconds and the bottom 32 bits are the
  // fractional part.
  // See RFC 1305 (NTP) and RFC 2030 (SNTP).
  uint64_t getTime(int index) const;

  // Gets the value at the given index only if it exists and is a 64-bit time.
  // This will return `true` if the value was successfully retrieved and
  // `false` otherwise.
  //
  // This function is useful to avoid two calls to `isTime`, once by the calling
  // code, and a second time by `getTime`.
  bool getIfTime(int index, uint64_t *v) const;

  // Gets the 64-bit double at the given index. This will return zero
  // if the index is out of range or if the argument is the wrong type.
  //
  // This will also return zero if the size of type `double` is not 8 bytes.
  double getDouble(int index) const;

  // Gets the value at the given index only if it exists and is a 64-bit double.
  // This will return `true` if the value was successfully retrieved and
  // `false` otherwise.
  //
  // This will also return false if the size of type `double` is not 8 bytes.
  //
  // This function is useful to avoid two calls to `isDouble`, once by the
  // calling code, and a second time by `getDouble`.
  bool getIfDouble(int index, double *v) const;

  // Gets the 32-bit char at the given index. This will return zero
  // if the index is out of range or if the argument is the wrong type.
  int32_t getChar(int index) const;

  // Gets the value at the given index only if it exists and is a 32-bit char.
  // This will return `true` if the value was successfully retrieved and
  // `false` otherwise.
  //
  // This function is useful to avoid two calls to `isChar`, once by the calling
  // code, and a second time by `getChar`.
  bool getIfChar(int index, int32_t *v) const;

  // Gets the boolean value at the given index. This will return false
  // if the index is out of range or if the argument is the wrong type.
  bool getBoolean(int index) const;

  // Gets the value at the given index only if it exists and is a boolean. This
  // This will return `true` if the value was successfully retrieved and
  // `false` otherwise.
  //
  // This function is useful to avoid two calls to `isBoolean`, once by the
  // calling code, and a second time by `getBoolean`.
  bool getIfBoolean(int index, bool *v) const;

 private:
  // Ensures that we have enough buffer capacity. This returns whether
  // we do, allocating if necessary. If there isn't enough space then
  // the memory error condition will be set to 'true'.
  //
  // This does not ensure size is the next multiple of 4.
  bool ensureCapacity(int size);

  // Ensures that we have enough capacity for the argument indexes. This
  // returns whether we do, allocating if necessary. If there isn't enough
  // space then the memory error condition will be set to 'true'.
  bool ensureArgIndexesCapacity(int size);

  // Aligns the given number to a multiple of 4.
  static int align(int n) {
    return ((n + 3) >> 2) << 2;
    // return off + ((4 - (off & 0x03)) & 0x03);
  }

  // Adds one argument having the specified size to the buffer. Things
  // are shifted around appropriately. This returns whether the addition
  // was a success. As well, bufSize_ will be set to the correct total size.
  //
  // argSize is not expected to be a multiple of 4, but it is sized
  // internally so that it is. Additionally, any extra allocated bytes
  // are set to zero.
  bool addArg(char tag, int argSize);

  // Parses a string starting at 'off' and returns the index just past
  // the NULL terminator. This will return a negative value if the end
  // of the string could not be found.
  static int parseString(const uint8_t *buf, int off, int len);

  // Parses all the arguments starting at 'off' and returns the index
  // just past the NULL terminator. This will return a negative value
  // if an error is encountered.
  int parseArgs(const uint8_t *buf, int off, int len);

  // Checks if the index is in range and the tag matches.
  bool isTag(int index, char tag) const {
    return (0 <= index && index < tagsLen_ - 1) &&
           buf_[tagsIndex_ + index + 1] == tag;
  }

  // Gets a big-endian-encoded uint32 from the given buffer.
  static uint32_t getUint(const uint8_t *buf) {
    return uint32_t{buf[0]} << 24 |
           uint32_t{buf[1]} << 16 |
           uint32_t{buf[2]} << 8 |
           uint32_t{buf[3]};
  }

  // Stores a big-endian-encoded uint32 into the given buffer.
  static void setUint(uint8_t *buf, uint32_t i) {
    buf[0] = i >> 24;
    buf[1] = i >> 16;
    buf[2] = i >> 8;
    buf[3] = i;
  }

  // Gets a big-endian-encoded uint64 from the given buffer.
  static uint64_t getUlong(const uint8_t *buf) {
    return (uint64_t{getUint(buf)} << 32) | uint64_t{getUint(buf + 4)};
  }

  // Stores a big-endian-encoded uint64 into the given buffer.
  static void setUlong(uint8_t *buf, uint64_t h) {
    setUint(buf, h >> 32);
    setUint(buf + 4, h);
  }

  // Similar to strcmp, this returns whether the two strings match, but
  // if they do not match, then loc is set to the location of the first
  // mismatch.
  // Named in honour of the naming of yore.
  static bool strcmploc(const char *s1, const char *s2, int *loc);

  // Buffer for holding the message
  uint8_t *buf_;
  int bufSize_;  // Invariant: bufSize_ % 4 == 0
  int bufCapacity_;
  bool dynamicBuf_;
  bool memoryErr_;

  // All the parts
  int addressLen_;
  int tagsIndex_;  // Invariant: tagsIndex_ % 4 == 0
  int tagsLen_;  // Includes the ',' if there are tags, zero otherwise
  int dataIndex_;  // Invariant: dataIndex_ % 4 == 0

  // Args
  int *argIndexes_;
  int argIndexesCapacity_;
  bool dynamicArgIndexes_;
};

// OSCBundle is a container for OSC messages and other OSC bundles.
// The internal buffer can be either dynamically allocated or set to a
// specific size. Any functions that add to, initialize, or change the
// bundle return whether they were successful. One of the possible
// reasons for failure is that there's not enough space in one of the
// internal buffers, and more cannot be allocated. The isMemoryError()
// function can be used to determine this case.
class OSCBundle {
 public:
  // Creates a new OSCBundle. The bundle time and maximum buffer size
  // are given. If the maximum buffer size is positive and less than 16,
  // then it will be set to 16. If it is non-positive, however, then the
  // buffer will be dynamically allocated as needed. It should be a
  // multiple of four.
  OSCBundle(int bufCapacity);

  // Creates a new OSCBundle that has a dynamic buffer.
  OSCBundle() : OSCBundle(0) {}

  // Not copyable
  OSCBundle(const OSCBundle &) = delete;
  OSCBundle &operator=(const OSCBundle &) = delete;

  ~OSCBundle();

  // Resets the contents and uses the given time for the bundle.
  // This also resets the memory error condition.
  //
  // This must be called at least once before messages or bundles can
  // be added.
  bool init(uint64_t time);

  // Returns the total size of the bundle.
  int size() const {
    return bufSize_;
  }

  // Gets a pointer to the internal buffer, for sending the data elsewhere.
  // See size() to retrieve the size. Play nice with this pointer.
  const uint8_t *buf() const {
    return buf_;
  }

  // Returns whether an insufficient buffer size is preventing the latest
  // content from being added.
  bool isMemoryError() const {
    return memoryErr_;
  }

  // Adds an OSC message to this bundle. This will return false if
  // the message could not be added, either due to init not being
  // called at least once or insufficient memory.
  bool addMessage(const LiteOSCParser &osc);

  // Adds another OSC bundle to this bundle. This will return false if
  // the message could not be added, either due to init not being called
  // at least once or insufficient memory.
  bool addBundle(const OSCBundle &bundle);

  // Parses the given buffer and returns whether it is a valid bundle. This
  // recursively looks into sub-bundles. If a bundle element is an OSC message
  // then a rudimentary check for starting with a '/' character is performed.
  static bool parse(const uint8_t *buf, int32_t len);

 private:
  // Adds content to the bundle. This returns false if init has not
  // been called at least once.
  bool add(const uint8_t *buf, int32_t size);

  // Ensures that we have enough buffer capacity. This returns whether
  // we do, allocating if necessary. If there isn't enough space then
  // the memory error condition will be set to 'true'.
  bool ensureCapacity(int size);

  uint8_t *buf_;
  int bufSize_;
  int bufCapacity_;
  bool dynamicBuf_;
  bool memoryErr_;

  bool isInitted_;
};

}  // namespace osc
}  // namespace qindesign

#endif  // LITEOSCPARSER_H_
